File: java/leetcode/src/main/java/com/leetcode/BFS_DFS/Combinations_77.java
package com.leetcode.BFS_DFS;

import java.util.ArrayList;
import java.util.List;

public class Combinations_77 {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        List<Integer> path = new ArrayList<>();
        dfs(n, 0, 1, k, path);
        return res;
    }
    void dfs(int n, int level, int start, int k, List<Integer> path) {
        if (level == k) {
            res.add(new ArrayList<>(path));
        } else {
            for(int i = start; i <= n; i++) {
                path.add(i);
                dfs(n, level + 1, i + 1, k, path);
                path.remove(path.size() - 1);
            }
        }
    }
}

File: js/algrithm/bfs-dfs/Combinations_77.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    const ans = []
    const dfs = function(start, level, path) {
        if (level === k) {
            ans.push(path.slice())
        } else {
            for (let i = start; i <= n; i++) {
                path.push(i)
                dfs( i + 1, level + 1, path)
                path.pop()
            }
        }
    }
    dfs(1, 0, [])
    return ans
};